TypeCobol/Compiler/Domain/SymbolTableBuilder.cs

```diff
@@ -0,0 +1,222 @@
+﻿using System;
+using TypeCobol.Compiler.CodeModel;
+using TypeCobol.Compiler.CupParser.NodeBuilder;
+using TypeCobol.Compiler.Nodes;
+using TypeCobol.Compiler.Parser;
+using TypeCobol.Compiler.Scopes;
+using TypeCobol.Compiler.Symbols;
+using TypeCobol.Tools.Options_Config;
+
+namespace TypeCobol.Compiler.Domain
+{
+    /// <summary>
+    /// Abstract base Class use to build the Symbol Table from a program perspective;
+    /// </summary>
+    public abstract class SymbolTableBuilder : ProgramClassBuilderNodeListener
+    {
+        /// <summary>
+        /// The abstract scope of the SymbolTable built.
+        /// </summary>
+        public abstract AbstractScope Scope
+        {
+            get;
+        }
+
+        /// <summary>
+        /// A Type Cobol configuration to be used.
+        /// </summary>
+        public static TypeCobolConfiguration Config
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Path of the Intrinsic file
+        /// </summary>
+        public static string IntrinsicPath
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// Called when A node has been syntactically recognized by the TypeCobol Parser.
+        /// </summary>
+        /// <param name="node">The node being built</param>
+        /// <param name="program">The Program that contains the node.</param>
+        public abstract override void OnNode(Node node, Program program);
+
+        /// <summary>
+        /// The Root Symbol Table.
+        /// </summary>
+#if DOMAIN_CHECKER
+        internal static RootSymbolTable _rootSymbolTable;
+#else
+        private static RootSymbolTable _rootSymbolTable;
+#endif
+
+        /// <summary>
+        /// Lock variable on the Rootymboltable access
+        /// </summary>
+        private static object rootLock = new object();
+
+        /// <summary>
+        /// Check if this Symbol table Builder has a Root Symbol Table.
+        /// </summary>
+        public static bool HasRoot
+        {
+            get
+            {
+                lock (rootLock)
+                {
+                    return _rootSymbolTable != null;
+                }
+            }
+        }
+        /// <summary>
+        /// The Root Symbol Table
+        /// </summary>
+        public static RootSymbolTable Root
```

Comment:
As discussed together , the RootSymbolTable cannot be static.
Currently the SymbolTable is scoped to a Snaphot of Node.
It means that each time a new parsing occurs, the SymbolTable is created from scratch and filled with data.

With RTC integration we'll see if we need to share SymbolTable accross CompilationProject, but I propose to keep the behavior identical to the current SymbolTable for this PR.

The only special behavior is for Intrinsic and dependencies. These are parsed only once and the result of the parsing is cloned into each "root" SymbolTable (SymbolTable of scope Intrinsic and Namespace in the current implementation) for each parsing.

Reply:
Ok done